### PR TITLE
Feature: Remember sort per folder

### LINKED ISSUE #63: Remember sort per folder
Sort column and direction is not currently persisted, and when set, stays across all folders until changed again. This enhancement would make Koffee's sort work more like Windows Explorer: treat each folder as having its own sort setting and persist it.

Add a `PathSort` property to History as a tuple of Path, Sort, bool (descending). When changing the sort by anything other than Name ascending, add it to the list. When sorting by Name ascending, remove it from the list since that is the default and to keep the list size down.

### PR DESCRIPTION
**Pull request by jilleJr**
Now it remembers the sorting per folder.

What it accomplishes:

- [x] Each folder has separate sorting
- [x] Only saves new sortings when `sort != (Name, false)`
- [x] Removes old sortings from save file when `sort = (Name, false)`
- [x] Tests to confirm this behaviour

I took some liberties in developing this and went a little out-of-spec. Fully OK to reject some of my ideas, and I'll revert them ASAP. The main differences are:

- Added a new type `PathSort` to represent a folders sorting. I did not replace the existing uses of `(SortField, bool)` tuples, but that might be something to consider for consistency (and more type-safety). I really dislike having unnamed fields, such as the unnamed bool.

- Stores the sort in a `Map<Path, PathSort>` instead of a `(Path, SortField, bool) list`. Mainly for improved search abilities, and the JSON looks nicer :)

- Added code to `Path` to make it serializable as dictionary keys and compatible with using as keys in `Map<'K, 'V>`'s

- Refactored a little in `MainController.fs`. Thought it was a little difficult to understand some of the code so I extracted some snippets into dedicated functions. Tried to keep my changes scoped to this PR but some code I couldn't refuse to touch.

## Preview

koffee-save-sort

The history.json file from the GIF above ends like this:

```json
  "PathSort": {
    "C:\\Users\\kalle": {
      "Sort": "Size",
      "Descending": true
    },
    "C:\\Users\\kalle\\Documents": {
      "Sort": "Modified",
      "Descending": true
    },
    "C:\\Users\\kalle\\Documents\\Programs": {
      "Sort": "Name",
      "Descending": true
    }
  }
}
```

This was an experience! It was fun to code more officially in F#. Still a big noob in when it comes to common F# practices and thinking habits. Will probably look into other issues on your repo here :)

---

Closes #63

### CODE REVIEW

**Comment by mattstermiller** [comment]
I scanned through the changes and I have to say I'm _very_ impressed with this as a first-time F# contribution, especially with the refactors you made to make the existing code easier to understand. I agree with your new PathSort type, that makes a lot of sense. I will go through it in-depth now.

**Comment by jilleJr** [comment]
Thank you! ^^

File: src/Koffee/MainController.fs [hunk 1]
```diff
@@ -43,6 +43,31 @@ let clearSearchProps model =
...
+    match path.NetHost with
+    | Some host -> hist.WithNetHost host
+    | None -> hist
+
+let private withBackStackedLocation model path =
+    if path <> model.Location then
+        { model.WithLocation path with
+            BackStack = (model.Location, model.Cursor) :: model.BackStack
+            ForwardStack = []
+            Cursor = 0
```
Comment (mattstermiller) [inline src/Koffee/MainController.fs:67]: The cursor being set here could be a bit surprising given the name of this function. Please move this line to after 102 with those properties

**Review by mattstermiller — CHANGES_REQUESTED** [review]
Outstanding PR. I have a few minor requests. In addition, please add the following line to RELEASE_NOTES.md with the existing "Adds" lines:
`- #92 Adds per-folder storage of sort settings (Thanks [your name or handle])`

File: src/Koffee/MainController.fs [hunk 2]
```diff
@@ -43,6 +43,31 @@ let clearSearchProps model =
...
+        |> List.map (fun path -> Item.Basic path path.Name NetHost)
+        |> Ok
+    else
+        fsReader.GetItems path |> actionError "open path"
+
+let private historyWithPathAndItsNetHost (history: History) path =
+    let hist = history.WithPath path
+    match path.NetHost with
+    | Some host -> hist.WithNetHost host
+    | None -> hist
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply src/Koffee/MainController.fs]: Also moved withBackStackedLocation into the MainModel type

File: src/Koffee/MainController.fs (hunk 1 above)
Comment (jilleJr) [reply src/Koffee/MainController.fs]: Fully agree. Moved

File: src/Koffee.Tests/PathTests.fs [hunk 3]
```diff
@@ -118,3 +119,45 @@ let ``Parent returns expected value`` pathStr expected =
...
+[<TestCase(@"/c/Sample2", @"/c/Sample1", 1)>]
+[<TestCase(@"/c/SAMPLE", @"/c/sample", 0)>]
+let ``Compare Unix paths`` aPath bPath expected =
+    testComparability aPath bPath expected
+
+[<TestCase(@"C:/Sample", @"/c/Sample", 0)>]
+[<TestCase(@"C:/Sample2", @"/c/Sample1", 1)>]
+[<TestCase(@"C:/SAMPLE", @"/c/sample", 0)>]
+let ``Compare Windows and Unix paths`` aPath bPath expected =
+    testComparability aPath bPath expected
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply src/Koffee.Tests/PathTests.fs]: Ah, my understanding was that the `Path.Format` did all the work, but now I see that `Path.Parse` normalizes the string and then `Path.Format` only alters it again if the target is Windows-style.

Updated accordingly.

File: src/Koffee.Tests/NavTests_History.fs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        }.WithLocation location
+    let fs = createFs()
+
+    // Act
+    let result = Nav.openPath fs newLocation SelectNone model
+
+    // Assert
+    let resultModel = expectOk "Failed to open path" result
+    resultModel.Sort |> shouldEqual (Some expected)
+
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply src/Koffee.Tests/NavTests_History.fs:119]: Thanks! Hope you don't mind the Arrange-Act-Assert convention I'm introducting. I really like that style and it goes by automatically by now. It works well in C#, but to repeat myself, I don't know enough F# conventions yet so I'm doing what I know.

File: src/Koffee.Tests/NavTests_History.fs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        folder "programs" [
+            file "koffee.exe"
+            file "notepad.exe"
+        ]
+        file "readme.md"
+    ]
+
+let expectOk onError = function
+    | Error _ -> failwith onError
+    | Ok result -> result
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply src/Koffee.Tests/NavTests_History.fs]: Nice catch!

File: RELEASE_NOTES.md [hunk 6]
```diff
@@ -3,6 +3,7 @@
 ### 0.6.0  Multi-select (WIP)
 - Changes UI to use new dark theme
 - Adds "Open File and Exit" command
+- #92 Adds per-folder storage of sort settings (Thanks @jilleJr)
```
Comment (jilleJr) [inline RELEASE_NOTES.md:6]: @mattstermiller  These are a little out of order right now. Should I move the first "Changes" line to below the "Adds" lines, or move the "Adds" note for this PR to at the end?

Comment (mattstermiller) [reply RELEASE_NOTES.md:6]: I try to arrange release notes by kind in this order:
1. major additions/changes
2. new features
3. minor improvement or changes
4. bug fixes

As such, it's in order as the latest new feature.

File: src/Koffee.Tests/NavTests_History.fs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattstermiller) [reply src/Koffee.Tests/NavTests_History.fs:119]: I've written all my tests by the AAA convention. I didn't include explicit comments them marking as such, but you may notice that my tests have 3 paragraphs ;)

**Comment by mattstermiller** [comment]
Thanks for your work on this!

**Pull request merged**